In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('adult.csv')

In [17]:
print("Info:")
df.info()
print("Head:")
display(df.head())
print("Null Sums:")
display(df.isnull().sum())




Info:
<class 'pandas.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   age             32561 non-null  int64
 1   workclass       32561 non-null  str  
 2   fnlwgt          32561 non-null  int64
 3   education       32561 non-null  str  
 4   education.num   32561 non-null  int64
 5   marital.status  32561 non-null  str  
 6   occupation      32561 non-null  str  
 7   relationship    32561 non-null  str  
 8   race            32561 non-null  str  
 9   sex             32561 non-null  str  
 10  capital.gain    32561 non-null  int64
 11  capital.loss    32561 non-null  int64
 12  hours.per.week  32561 non-null  int64
 13  native.country  32561 non-null  str  
 14  income          32561 non-null  str  
dtypes: int64(6), str(9)
memory usage: 3.7 MB
Head:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


Null Sums:


age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64

## TASK 1

In [29]:
df.replace('?', np.nan, inplace=True)
display(df.isnull().sum())

age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64

## TASK 2

In [31]:
missing_count = df.isnull().sum()
missing_percentage = (df.isnull().mean() * 100)

# Combine count and percentage into one table
missing_summary = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing Percentage': missing_percentage
})

missing_summary = missing_summary.sort_values(by='Missing Count', ascending=False)

missing_summary

,Missing Count,Missing Percentage
occupation,1843,5.660146
workclass,1836,5.638647
native.country,583,1.790486
fnlwgt,0,0.000000
education,0,0.000000
education.num,0,0.000000
age,0,0.000000
marital.status,0,0.000000
relationship,0,0.000000
sex,0,0.000000


## TASK 3

For "workingclass" and "occuptaion" we chose "unknown" because for categorical data, "unknown" is a reasonable choice because the missing value doesn't necessarily mean that the person belongs to the most common category. Using the mode could incorrectly assign missing people to the most frequent category.

For numerical data we would choose mode if the data was more normal and median if there were larger outliers

In [38]:
df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')

df[['workclass', 'occupation']].isnull().sum()


workclass     0
occupation    0
dtype: int64

## TASK 4

In [39]:
exact_duplicates = df.duplicated().sum()

print("Number of exact duplicate rows:", exact_duplicates)

Number of exact duplicate rows: 24


In [47]:
duplicate_rows = df[df.duplicated(keep=False)]

duplicate_rows

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
6227,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
7615,19,Private,251579,Some-college,10,Never-married,Other-service,Own-child,White,Male,0,0,14,United-States,<=50K
7978,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
8356,21,Private,250051,Some-college,10,Never-married,Prof-specialty,Own-child,White,Female,0,0,10,United-States,<=50K
8453,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
8500,38,Private,207202,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,48,United-States,>50K
8645,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
9067,27,Private,255582,HS-grad,9,Never-married,Machine-op-inspct,Not-in-family,White,Female,0,0,40,United-States,<=50K
9269,20,Private,107658,Some-college,10,Never-married,Tech-support,Not-in-family,White,Female,0,0,10,United-States,<=50K
9305,25,Private,195994,1st-4th,2,Never-married,Priv-house-serv,Not-in-family,White,Female,0,0,40,Guatemala,<=50K


In [49]:
#Remove the target column for feature-level duplicate checking
features_only = df.drop(columns=['income'])

#Count duplicates ignoring the income label
duplicates_without_target = features_only.duplicated().sum()

print(
    "\nDuplicates when ignoring income label:",
    duplicates_without_target
)

#Display feature-level duplicates
duplicate_features = df[features_only.duplicated(keep=False)]

print("\nRows duplicated when ignoring income:")
display(duplicate_features)


Duplicates when ignoring income label: 25

Rows duplicated when ignoring income:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
4963,39,Private,138192,Bachelors,13,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,40,United-States,<=50K
6227,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
7615,19,Private,251579,Some-college,10,Never-married,Other-service,Own-child,White,Male,0,0,14,United-States,<=50K
7978,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
8356,21,Private,250051,Some-college,10,Never-married,Prof-specialty,Own-child,White,Female,0,0,10,United-States,<=50K
8453,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
8500,38,Private,207202,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,48,United-States,>50K
8645,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
9067,27,Private,255582,HS-grad,9,Never-married,Machine-op-inspct,Not-in-family,White,Female,0,0,40,United-States,<=50K
9269,20,Private,107658,Some-college,10,Never-married,Tech-support,Not-in-family,White,Female,0,0,10,United-States,<=50K


## TASK 5

In [54]:
print("Education:")
display(df['education'].unique())

print("\nMarital Status:")
display(df['marital.status'].unique())

print("\nNative Country:")
display(df['native.country'].unique())

Education:


<StringArray>
[     'HS-grad', 'Some-college',      '7th-8th',         '10th',
    'Doctorate',  'Prof-school',    'Bachelors',      'Masters',
         '11th',   'Assoc-acdm',    'Assoc-voc',      '1st-4th',
      '5th-6th',         '12th',          '9th',    'Preschool']
Length: 16, dtype: str


Marital Status:


<StringArray>
[              'Widowed',              'Divorced',             'Separated',
         'Never-married',    'Married-civ-spouse', 'Married-spouse-absent',
     'Married-AF-spouse']
Length: 7, dtype: str


Native Country:


<StringArray>
[             'United-States',                          nan,
                     'Mexico',                     'Greece',
                    'Vietnam',                      'China',
                     'Taiwan',                      'India',
                'Philippines',            'Trinadad&Tobago',
                     'Canada',                      'South',
         'Holand-Netherlands',                'Puerto-Rico',
                     'Poland',                       'Iran',
                    'England',                    'Germany',
                      'Italy',                      'Japan',
                       'Hong',                   'Honduras',
                       'Cuba',                    'Ireland',
                   'Cambodia',                       'Peru',
                  'Nicaragua',         'Dominican-Republic',
                      'Haiti',                'El-Salvador',
                    'Hungary',                   'Columbia',
          

In [63]:

df['education'] = df['education'].str.strip()
df['marital.status'] = df['marital.status'].str.strip()
df['native.country'] = df['native.country'].str.strip()

print("\nEducation after cleaning:")
print(df['education'].unique())

print("\nMarital Status after cleaning:")
print(df['marital.status'].unique())

print("\nNative Country after cleaning:")
print(df['native.country'].unique())


Education after cleaning:
<StringArray>
[     'HS-grad', 'Some-college',      '7th-8th',         '10th',
    'Doctorate',  'Prof-school',    'Bachelors',      'Masters',
         '11th',   'Assoc-acdm',    'Assoc-voc',      '1st-4th',
      '5th-6th',         '12th',          '9th',    'Preschool']
Length: 16, dtype: str

Marital Status after cleaning:
<StringArray>
[              'Widowed',              'Divorced',             'Separated',
         'Never-married',    'Married-civ-spouse', 'Married-spouse-absent',
     'Married-AF-spouse']
Length: 7, dtype: str

Native Country after cleaning:
<StringArray>
[             'United-States',                          nan,
                     'Mexico',                     'Greece',
                    'Vietnam',                      'China',
                     'Taiwan',                      'India',
                'Philippines',            'Trinadad&Tobago',
                     'Canada',                      'South',
         'Holand-N

## TASK 6

In [65]:
# Numeric descriptive statistics
print("Numeric Descriptive Statistics:")
display(df.describe())


Numeric Descriptive Statistics:


,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [68]:

# Categorical descriptive statistics
print("\nCategorical Descriptive Statistics:")
display(df.describe(include='object'))


Categorical Descriptive Statistics:


C:\Users\STONE\AppData\Local\Temp\ipykernel_17200\1767931716.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  display(df.describe(include='object'))


,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,31978,32561
unique,9,16,7,15,6,5,2,41,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


In [71]:

# Age
age_mean = df['age'].mean()
age_median = df['age'].median()

print("\nAge:")
print("Mean:", age_mean)
print("Median:", age_median)

# Hours per week
hours_mean = df['hours.per.week'].mean()
hours_median = df['hours.per.week'].median()

print("\nHours-per-week:")
print("Mean:", hours_mean)
print("Median:", hours_median)

# Differences
print("\nDifference between mean and median:")
print("Age:", abs(age_mean - age_median))
print("Hours-per-week:", abs(hours_mean - hours_median))


Age:
Mean: 38.58164675532078
Median: 37.0

Hours-per-week:
Mean: 40.437455852092995
Median: 40.0

Difference between mean and median:
Age: 1.581646755320783
Hours-per-week: 0.4374558520929952


The mean age is 38.58 years and the median age is 37 years, giving a difference of approximately 1.58 years. The mean hours-per-week is 40.44 hours while the median is 40 hours, giving a difference of only 0.44 hours. Overall, the mean and median are relatively close for both variables, suggesting that neither distribution differs dramatically between its average and central value.

## TASK 7

In [72]:
occupation_counts = df['occupation'].value_counts()

print("=== Occupation Frequency ===")
print(occupation_counts)

print("\nMost common occupation:", occupation_counts.idxmax())
print("Frequency:", occupation_counts.max())


=== Occupation Frequency ===
occupation
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Unknown              1843
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Name: count, dtype: int64

Most common occupation: Prof-specialty
Frequency: 4140


In [73]:
sex_percentage = df['sex'].value_counts(normalize=True) * 100

print("\n=== Sex Percentage Distribution ===")
print(sex_percentage)



=== Sex Percentage Distribution ===
sex
Male      66.920549
Female    33.079451
Name: proportion, dtype: float64


In [74]:

income_counts = df['income'].value_counts()
income_percentage = df['income'].value_counts(normalize=True) * 100

print("\n=== Income Counts ===")
print(income_counts)

print("\n=== Income Percentage Distribution (IMBALANCED) ===")
print(income_percentage)


=== Income Counts ===
income
<=50K    24720
>50K      7841
Name: count, dtype: int64

=== Income Percentage Distribution ===
income
<=50K    75.919044
>50K     24.080956
Name: proportion, dtype: float64


## TASK 8

In [76]:
# clean
df['education'] = df['education'].str.strip()

# Check mapping from education label to numerical value
edu_mapping = df.groupby('education')['education.num'].unique()
print(edu_mapping)

# Check reverse mapping from numerical value to education label
reverse_mapping = df.groupby('education.num')['education'].unique()
print(reverse_mapping)

education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object
education.num
1        [Preschool]
2          [1st-4th]
3          [5th-6th]
4          [7th-8th]
5              [9th]
6             [10th]
7             [11th]
8             [12th]
9          [HS-grad]
10    [Some-college]
11       [Assoc-voc]
12      [Assoc-acdm]
13       [Bachelors]
14         [Masters]
15     [Prof-school]
16       [Doctorate]
Name: education, dtype: object


## TASK 9

In [79]:
import sqlite3
import pandas as pd

df_raw = pd.read_csv('adult.csv')

conn = sqlite3.connect('adult_data.db')
df_raw.to_sql('adult_income', conn, index=False, if_exists='replace')

query = "SELECT * FROM adult_income WHERE age > 30;"
df_sql = pd.read_sql_query(query, conn)

conn.close()

print(f"Original CSV shape: {df_raw.shape}")
print(f"SQL Extracted shape: {df_sql.shape}")

Original CSV shape: (32561, 15)
SQL Extracted shape: (21989, 15)


## TASK 10

### Dataset OverviewDimensions: 
32,561 rows and 15 columns.  Representation: Demographics, employment metadata, and financial metrics extracted from the 1994 US Census dataset to predict individual income class ( <=50K vs >50K ).  

### Data Quality Issues FoundHidden Missing Values:
Initially hidden as '?' strings across categorical fields:  workclass: 1,836 missing values (5.64%)  occupation: 1,843 missing values (5.66%)  native.country: 583 missing values (1.79%)  Duplicates: 24 exact duplicate rows detected across all attributes.  Inconsistent Categorical Data: Strings in categorical columns (education, marital.status, native.country, etc.) contained leading and trailing whitespace.  

### Key ObservationsClass Imbalance:
The target variable income is heavily imbalanced, with 75.92% earning <=50K and 24.08% earning >50K.  Most Common Occupation: Prof-specialty is the single most frequent occupation, representing 12.71% of the dataset.  Feature Redundancy: education and education.num maintain a perfect 1-to-1 deterministic relationship. 

### Cleaning Decisions Made:
Whitespace Trimming: Applied .str.strip() to eliminate leading/trailing whitespace across all text columns.  Missing Value Standardization: Replaced hidden '?' markers with standard np.nan values.  Missing Value Handling: Mode/Unknown category imputation is recommended for workclass, occupation, and native.country to preserve sample size.  Duplicate Removal: Removed 24 exact duplicate records to prevent skewed frequency weights.